Replace torch.export default decomp table to be lazily populated

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137650

In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

tugsbayasgalan committed:
TEMP

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).

tugsbayasgalan committed:
Update on "TEMP"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -681,6 +677,197 @@ def _decompose_exported_program(
     return exported_program
 
 
+class CustomDecompTable(Dict[torch._ops.OperatorBase, Callable]):
```
maybe this class implementation should live in a private file, so it isn't part of the `torch.export.exported_program` public module?

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -210,6 +210,7 @@ def register(op):
 def get_decompositions(
     aten_ops: Sequence[Union[torch._ops.OperatorBase, OpOverloadPacket]],
     type: str = "post_autograd",
+    ignore_cia: bool = False,
```
it feels a bit weird that the export lazy dict logic needs to leak into this function. Why do we need it? (vs. just having the lazy dict get all ATen CIA ops the first time)

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

bdhirsh commented on torch/export/decomp_utils.py:
```diff
@@ -0,0 +1,187 @@
+    until we really need to materialize it (which is when we run decomposition pass.)
+
+    Invariant we hold is:
+    1. If it is read operation on this dict, we always materialize
```
Maybe a bit more concretely:

(1) all ATen decomps are loaded into the dict on creation

(2) custom op decomps are loaded lazily the first time the dictionary is ever read from (we delay loading to make it more likely that any user custom ops will have been loaded at this point)

(3) we also load a final time during export, right before calling `run_decompositions()`

bdhirsh commented on torch/export/decomp_utils.py:
```diff
@@ -0,0 +1,187 @@
+        if key in self.aten_decomp_table:
+            return self.aten_decomp_table[key]
+
+        if _is_aten_op(key):
```
This is unused inside of this function?

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -263,170 +263,51 @@ def remove_decompositions(
-        return decomp_table
+        decomp_table_for_functional_cia_ops_legacy = get_decompositions(
+            [
+                aten.all.dimname,
```
Can you help me understand the migration story here: it looks like prior to this PR, this extra list of decomps used to be in the `not config.is_fbcode()` path, but now it is flipped (it only shows up in the `is_fbcode()` path)? Or am I mis-reading

(also, it looks like you dropped the comment associated with the list on like 399)

tugsbayasgalan replied:
So what's happening here is:
1. The clean list without any functional CIA is in _core_aten_decompositions_post_autograd(). 
2. Previously, if it is fbcode, we would try to return the decomp table without any functional CIA ops due to my changes like 3 weeks ago. But in this PR, I am trying to go WAY BACK to make sure the core_aten_decomp has the entries for those ignored functional CIA as well. I think it makes no difference tho in practice

I will add the comment back haha

bdhirsh replied:
(agreed offline to remove the extra decomps)

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -437,8 +318,13 @@ def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
 # to their default decomp. In old export, this will
 # be decomposed implicitly.
 def _decomp_table_to_post_autograd_aten():
+    from torch._export.utils import (
```
Let me know if this sounds right to you in terms of the long term state:

(1) we don't need this extraa `torch._decomp._decomp_table_to_post_autograd_aten` API at all

(2) `torch._decomp.core_aten_decompositions()` above becomes a one-liner that just returns `torch.export.default_decompositions()`

If that sounds right, then is all of the complexity in this file about CIA ops a temporary state until we migrate?

tugsbayasgalan replied:
Yep this is just a table to make sure we don't break other ppl as there are some users who are still relying on post-autograd export behavior. And yes for (2), `torch._decomp.core_aten_decompositions()` will be an alias to `torch.export.default_decompositions()` cc @avikchaudhuri

bdhirsh replied:
ok - in that case, looking forward to the cleanup PR :)

tugsbayasgalan replied:
Actually there are only like two places where this matters. I will just rewrite it there at their callsite.

bdhirsh commented on torch/_export/utils.py:
```diff
@@ -918,3 +922,161 @@ def _disable_load_state_dict_hooks(mod: torch.nn.Module):
+        if op_namespace_name != "aten":
+            cia_ops |= _collect_all_valid_cia_ops_for_namespace(op_namespace_name)
+        else:
+            cia_ops |= _collect_all_valid_cia_ops_for_aten_namespace()
```
oh I guess you have a separate helper for the aten case because the aten ops are safe to cache (since we generate them all at `import torch` time, which is not true for the custom op namespaces. probably worth a comment?

tugsbayasgalan replied:
Yep

bdhirsh commented on torch/export/__init__.py:
```diff
@@ -50,7 +50,7 @@
     "ExportedProgram",
     "ModuleCallEntry",
     "ModuleCallSignature",
-    "core_aten_decompositions",
```
oh I didn't realize `torch.export.core_aten_decompositions` was a public API. are we ok with fully deleting it in this PR?

tugsbayasgalan replied:
Yeah I added this a couple weeks ago, this diff just renames it.

bdhirsh commented on torch/export/_decomp_utils.py:
```diff
@@ -0,0 +1,178 @@
+        )
+
+    def __delitem__(self, key):
+        self.pop(key)
```
hmm is there any reason that we can't just inherit the existing impl from dict, since we are subclassing dict here? (`__delitem__` and `__missing__` don't seem like they need any special handling)

tugsbayasgalan replied:
Delitem needs to be special because we could have pending operators in self.deleted_custom_ops set.

bdhirsh commented on torch/export/_decomp_utils.py:
```diff
@@ -0,0 +1,178 @@
+        if _is_aten_op(key):
+            return key in self.aten_decomp_table
+
+        if key in self.additional_custom_op_decomp:
```
hmm up to you, but it does feel like:

(1) you need to implement quite a few dictionary methods due to needing to materialize

(2) they would all be a bit simpler to read if we have a single `self.decomp_table`, instead of separating the aten and custom op decomps (so it's not clear to me that it's a net complexity decrease to separate the two)

bdhirsh commented on torch/export/_decomp_utils.py:
```diff
@@ -0,0 +1,178 @@
+        )
+
+    def __reverse__(self):
+        raise RuntimeError("Cannot call reverse() on custom decomp table")
```
just curious, but why does the default reverse not just work?

tugsbayasgalan replied:
Ahh this is because we have two tables..

bdhirsh reviewed (commented):
sgtm!

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).

bdhirsh commented on test/export/test_export.py:
```diff
@@ -1756,6 +1753,71 @@ def forward(self, x):
+        # for op in all_ops:
+        #     if op in table:
+        #        del table[op]
+        with torch.library._scoped_library("mylib", "FRAGMENT") as lib:
```
Another nice test to have:

```
table = default_decompositions()
del table[torch.ops.aten.linear.default]

# define a new custom CIA op (proxy for loading a custom op library via import)

# export a program with that custom CIA op in it, confirm that it is still decomposed
ep = ep.run_decompositions(decomp_table)
```

tugsbayasgalan replied:
Added.

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).

tugsbayasgalan committed:
Update on "Replace torch.export default decomp table to be lazily populated"


In this PR, we implement lazy dictionary for export decomp behaviour for following reasons:
1. Custom op loading can happen after import time, as a result, the decomp table might not be able to pick up the decomp. Therefore we try to delay materialization as late as possible. 

I intentionally seperated out the core_aten_decomp to not have any custom CIA ops in this PR to mitigate the risk of getting reverted but in the future, core_aten_decomp under torch/_decomp will exist as an alias to official export table (torch.export.default_decompositions) 

Differential Revision: [D64140807](https://our.internmc.facebook.com/intern/diff/D64140807)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64140807).
